### PR TITLE
Paginate search results in InstantSearchScreen - 1

### PR DESCRIPTION
**Pull request by msasikanth**

### CODE REVIEW

**Comment by msasikanth** [comment]
Most of the additions in this PR are related to the copy of adapter-related classes like `PagingItemAdapter` & `InstantSearchResultItemType`. We had to duplicate and keep the older usage since only all patient's adapter is migrated and we want the search results adapter to work as is, in this PR.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt [hunk 1]
```diff
@@ -39,10 +39,11 @@ data class InstantSearchModel(
   companion object {
     fun create(
         additionalIdentifier: Identifier?,
-        patientPrefillInfo: PatientPrefillInfo?
+        patientPrefillInfo: PatientPrefillInfo?,
+        searchQuery: String?
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:43]: Is there a reason this change is part of this PR? 'Instant Search Improvements' is vague doesn't tell us what is actually happening in a particular changeset. Ideally, supporting the new way of pagination and support for prefilling search query should be two different PRs.

**Review by vinaysshenoy — CHANGES_REQUESTED** [review]
While the changes themselves look good (apart from some minor corrections), I think we can split this changeset into two different PRs for cleaner history. Can we do that please?

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:43]: I had to do this here mainly to avoid the load all patients method being called twice and causing issues with the paging source. Since we are not technically migrating from older paging implementation in this screen, I have decided to called it instant search screen improvements and make the necessary changes required for the instant search to work correctly. wdyt?

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchUpdate.kt [hunk 2]
```diff
@@ -171,11 +171,6 @@ class InstantSearchUpdate @Inject constructor(
   ): Next<InstantSearchModel, InstantSearchEffect> {
     if (model.hasSearchQuery) return noChange()
 
-    val effect = if (event.patients.isNotEmpty())
-      ShowAllPatients(event.patients, model.facility!!)
-    else
-      ShowNoPatientsInFacility(model.facility!!)
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchUpdate.kt:177]: For now yeah, we are going to use the load state listeners to call the empty views. For the future, I am making changes to use one single adapter that will let us move that logic to the Mobius loop. So basically we will trigger an event every time the load state listener updates and do the conditional check there and use UI renderer to show/hide the empty state.

File: app/src/test/java/org/simple/clinic/instantsearch/InstantSearchResultsItemTypeTest.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    // when
+    val header = InstantSearchResultsItemType.insertHeaders(
+        beforeSearchResult = searchResult1,
+        afterSearchResult = searchResult2,
+        currentFacility = currentFacility
+    )
+
+    // then
+    assertThat(header).isEqualTo(null)
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/test/java/org/simple/clinic/instantsearch/InstantSearchResultsItemTypeTest.kt:127]: Huh, for some reason the first time I tried that, it didn't auto complete. So I assumed there was no such method, it's working now.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:43]: We could make that change first and then switch to the new paging implementation, right? The two need not go together.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt [hunk 4]
```diff
@@ -101,8 +104,13 @@ class InstantSearchEffectHandler @AssistedInject constructor(
     return ObservableTransformer { effects ->
       effects
           .observeOn(schedulers.io())
-          .map { patientRepository.allPatientsInFacility_old(it.facility) }
-          .map(::AllPatientsLoaded)
+          .flatMap { (facility) ->
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt:110]: Ah, good point. We can get the latest effects instead of constructing the pager every time and then getting the latest pager.

File: app/src/androidTest/java/org/simple/clinic/PagingTestCase.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import kotlinx.coroutines.rx2.rxObservable
+
+class PagingTestCase<K : Any, V : Any>(
+    private val pagingSource: PagingSource<K, V>,
+    loadSize: Int,
+    key: K? = null,
+    placeholdersEnabled: Boolean = false
+) {
+
+  private val loadParams = Refresh(key, loadSize, placeholdersEnabled)
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/androidTest/java/org/simple/clinic/PagingTestCase.kt:16]: One of the reasons why I didn't do that is, we are not using the other sub-types. Basically, most of our scenarios just test whether or not we are fetching the data when we load this call and not actually testing paging, for example, we `Append` when the page reaches the end and we want to load more. 

But I can make the change if you want it for future scenarios, should I make the change?

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt [hunk 6]
```diff
@@ -101,8 +104,13 @@ class InstantSearchEffectHandler @AssistedInject constructor(
...
+            pagerFactory.createPager(
+                sourceFactory = { patientRepository.allPatientsInFacility(facilityId = facility.uuid) },
+                pageSize = instantSearchConfig.pagingLoadSize
+            )
+          }
+          .switchMap { Observable.just(AllPatientsLoaded(it)) }
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt:116]: Ah, should I name it `AllPatientsInFacilityLoaded` then?

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchResultsItemType.kt [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        currentFacility: Facility
+    ): InstantSearchResultsItemType? {
+      return when {
+        canAddAssignedFacility(beforeSearchResult, afterSearchResult) -> AssignedFacilityHeader(facilityName = currentFacility.name)
+        canAddNearbyFacilityHeader(beforeSearchResult, afterSearchResult) -> NearbyFacilitiesHeader
+        else -> null
+      }
+    }
+
+    private fun canAddAssignedFacility(beforeSearchResult: SearchResult?, afterSearchResult: SearchResult?) =
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchResultsItemType.kt:48]: Both seem kinda same to me tbh 😅 . I used `can` prefix mainly because we are calling this method in conditional.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:43]: Okay, let me break out these commits then.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchEffectHandler.kt:116]: Renamed it

File: app/src/androidTest/java/org/simple/clinic/PagingTestCase.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/PagingTestCase.kt:16]: Ah, that's fine then. Just wanted to verify this. We can make this change later if we need to.

**Review by vinaysshenoy — APPROVED** [review]
